### PR TITLE
#36 fix missing unit number for open statements

### LINKED ISSUE #36: fparser2: generate correct OPEN call when unit number argument is not named
When parsing the following OPEN call:

    OPEN( idrst, FILE = TRIM(cdname), FORM = 'unformatted', ACCESS = 'direct'   &
            &       , RECL = 8, STATUS = 'old', ACTION = 'read', IOSTAT = ios, ERR = 987 )

fparser2 generates:

    OPEN(UNIT = <class 'fparser.Fortran2003.File_Unit_Number'>, FILE = TRIM(cdname), FORM = 'unformatted', ACCESS = 'direct', RECL = 8, STATUS = 'old', ACTION = 'read', IOSTAT = ios, ERR = 987)

i.e. the fact that the unit number argument isn't named in the original call appears to cause problems.

### PR DESCRIPTION
**Pull request by arporter**
This PR fixes #36 by correcting the `match` method of `Connect_Spec`.

### CODE REVIEW

**Comment by arporter** [comment]
I've just tested this branch with the ORCA2_LIM configuration of NEMO. Modulo the problem with INCLUDE statements (they get removed), fparser2 successfully parses and re-generates the whole code base which I've then compiled (with the Intel compiler).

File: src/fparser/Fortran2003.py [hunk 1]
```diff
@@ -6028,8 +6035,11 @@ def match(string):
                 ('IOMSG', Iomsg_Variable),
                 ('FILE', File_Name_Expr),
                 ('UNIT', File_Unit_Number)]:
-            obj = KeywordValueBase.match(keyword, value, string,
-                                         upper_lhs=True)
+            try:
+                obj = KeywordValueBase.match(keyword, value, string,
+                                             upper_lhs=True)
+            except NoMatchError:
```
Comment (rupertford) [inline src/fparser/Fortran2003.py]: put back NoMatchError as explained in the issue. Good.

**Comment by arporter** [comment]
In trying to get test coverage for the missed lines highlighted in the review I discovered that the parser failed if a keyword is misspelt (e.g "aunit=23" instead of "unit=23"). I've fixed that bug and obtained coverage of the modified lines. It's probable that this bug will exist in other places in the code but I haven't checked. This possibly points to the need to have a KeywordValueList base class that would cut down on code duplication.

**Comment by arporter** [comment]
Brought branch up-to-date with master. Back to @rupertford for second review...

File: src/fparser/tests/fparser2/test_Fortran2003.py [hunk 2]
```diff
@@ -3664,42 +3786,43 @@ def test_Contains(): # R1237
...
-        use_names = obj.__dict__.get('use_names',None)
-        if not use_names: continue
-        match = obj.__dict__.get('match',None)
+    NOF_NEEDED_TESTS = 0
+    NOF_NEEDED_MATCH = 0
+    TOTAL_NEEDS = 0
+    TOTAL_CLASSES = 0
+    for NAME in dir():
+        OBJ = eval(NAME)
+        if not isinstance(OBJ, ClassType): continue
```
Comment (rupertford) [inline src/fparser/tests/fparser2/test_Fortran2003.py]: The if should be over multiple lines. I'll ignore as you were only changing the line to improve pep8/pylint.

File: src/fparser/tests/fparser2/test_Fortran2003.py [hunk 3]
```diff
@@ -3664,42 +3786,43 @@ def test_Contains(): # R1237
...
+        if not issubclass(OBJ, Base): continue
```
Comment (rupertford) [inline src/fparser/tests/fparser2/test_Fortran2003.py]: As above

File: src/fparser/tests/fparser2/test_Fortran2003.py [hunk 4]
```diff
@@ -3664,42 +3786,43 @@ def test_Contains(): # R1237
...
+        CLSNAME = OBJ.__name__
+        if CLSNAME.endswith('Base'): continue
```
Comment (rupertford) [inline src/fparser/tests/fparser2/test_Fortran2003.py]: As above

File: src/fparser/tests/fparser2/test_Fortran2003.py [hunk 5]
```diff
@@ -3664,42 +3786,43 @@ def test_Contains(): # R1237
...
+        TOTAL_CLASSES += 1
+        SUBCLASS_NAMES = OBJ.__dict__.get('subclass_names', None)
+        USE_NAMES = OBJ.__dict__.get('use_names', None)
+        if not USE_NAMES: continue
```
Comment (rupertford) [inline src/fparser/tests/fparser2/test_Fortran2003.py]: As above

**Pull request merged**